code/model/editableformfields/EditableFormField.php

```diff
@@ -335,20 +345,30 @@ public function canDelete($member = null) {
 	}
 
 	/**
-	 * Return whether a user can edit this form field
-	 * based on whether they can edit the page
+	 * Return whether a user can view this form field
+	 * based on whether they can view the page
 	 *
 	 * @return bool
 	 */
-	public function canEdit($member = null) {
+	public function canView($member = null) {
 		if($this->Parent()) {
-			return $this->Parent()->canEdit($member) && !$this->isReadonly();
+			return $this->Parent()->canView($member) && !$this->isReadonly();
 		}
 
 		return true;
 	}
 
 	/**
+	 * Return whether a user can create this form field
+	 * based on whether they can edit the page
+	 *
+	 * @return bool
+	 */
+	public function canCreate($member = null) {
```

Comment:
Context is the solution for "we need singletons to have knowledge about potential values for future properties during canCreate".

See https://github.com/silverstripe/silverstripe-cms/blob/3/code/model/SiteTree.php#L989 for how this has been implemented in SiteTree::canView()

At the moment you can only access that parameter via func_get_arg(1), but it'll be a standard parameter in 4.0.

The problem is that gridfield might not support it yet, so I need to do some digging and get back to you with a solution that works with userforms.